scipy/cluster/_hierarchy.pyx

```diff
@@ -914,16 +939,17 @@ cdef class LinkageUnionFind:
 
 
 cdef label(double[:, :] Z, int n):
+    """Correctly label clusters in unsorted dendrogram."""
     cdef LinkageUnionFind uf = LinkageUnionFind(n)
     cdef int i, x, y, x_root, y_root
     for i in range(n - 1):
         x, y = int(Z[i, 0]), int(Z[i, 1])
         x_root, y_root = uf.find(x), uf.find(y)
-        uf.merge(x, y)
         if x_root < y_root:
             Z[i, 0], Z[i, 1] = x_root, y_root
         else:
             Z[i, 0], Z[i, 1] = y_root, x_root
+        Z[i, 3] = uf.merge(x_root, y_root)
```

Comment:
This is a fix in ``label`` that affects ``nn_chain`` as well right? And the issue is covered by one of the new tests?